core/unix/signal.c

```diff
@@ -2227,18 +2362,41 @@ signal_swap_mask(dcontext_t *dcontext, bool to_app)
              * We need to remove our own init-time handler and mask.
              */
             unset_initial_crash_handlers(dcontext);
+            DOLOG(4, LOG_ASYNCH, dump_unmasked(dcontext, __FUNCTION__););
             return;
         }
+        d_r_mutex_lock(&info->sigblocked_lock);
         sigprocmask_syscall(SIG_SETMASK, &info->app_sigblocked, NULL,
                             sizeof(info->app_sigblocked));
+        d_r_mutex_unlock(&info->sigblocked_lock);
     } else {
+        /* Normally we get here just once at takeover, but if we have threads
+         * going native we don't consider them unmasked as we can't rely on their
+         * mask not changing: so we undo their counts when the come back here and
+         * redo with the latest blocked signals.
+         */
+        for (int i = 1; i <= MAX_SIGNUM; i++) {
+            if (!EMULATE_SIGMASK(info, i))
+                continue;
+            if (kernel_sigismember(&info->app_sigblocked, i))
+                ATOMIC_INC(int, info->sighand->threads_unmasked[i]);
+        }
+        d_r_mutex_lock(&info->sigblocked_lock);
         unblock_all_signals(&info->app_sigblocked);
```

Comment:
Can you explain this else-block, and the case when it is invoked? Why do we need to increment and then decrement the unmasked thread count.

Reply:
This line is setting app_sigblocked to a new absolute value.  The only way to update the counts is to first remove all the counts from our old value of app_sigblocked and then re-increment for the new values.  This is also how we do it for sigprocmask with SIG_SETMASK.  I added a sentence making it clear that unblock_all_signals *writes* to app_sigblocked.

Reply:
Thanks. I was thinking only about takeover, but from your comment I understand that this is also required for threads that were previously under DR control, went back native, and then again under DR control. That could be caused by multiple calls to dr_app_start I suppose. From code I see this can also happen on a "call to a native module"?

Reply:
There is a -native_exec feature which was used to run .NET and Java natively as a stopgap to get apps working that had just a little dynamic scripting.  It has lots of caveats of course.